runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClient.java

```diff
@@ -148,16 +148,14 @@ public void onNext(BeamFnApi.InstructionResponse response) {
       LOG.debug("Received InstructionResponse {}", response);
       CompletableFuture<BeamFnApi.InstructionResponse> responseFuture =
           outstandingRequests.remove(response.getInstructionId());
-      if (responseFuture != null) {
-        if (response.getError().isEmpty()) {
-          responseFuture.complete(response);
-        } else {
-          responseFuture.completeExceptionally(
-              new RuntimeException(
-                  String.format(
-                      "Error received from SDK harness for instruction %s: %s",
-                      response.getInstructionId(), response.getError())));
-        }
+      if (response.getError().isEmpty()) {
```

Comment:
I do not like this change, because of my comment below about NPE.

But also, I do not know the code well enough. Is there a distributed systems question? Can a duplicate message be received that should be recieved?

If the answer is "yes" then the previous code is correct. If the answer is "no" then the previous code should crash or at least log an error.

Reply:
Thanks for the quick reply!

Is there some logic for the resent the message?  I have not find it, so  I think in this case there is no duplicate message. 

Please correct me if there anything I misunderstand :)

Reply:
I read some code and thought about this. The message comes from the other side of a gRPC connection. We cannot assume what code is connected. It could be any of the SDK harnesses, or could be another SDK harness not part of Beam.

Receiving a response for a request where we are not waiting for an answer is OK. Maybe it means there is a bug in code on the other side of the connection. I don't think the runner side should crash and terminate. It may be worth a log message, but then a problem on the SDK harness could cause flooding on the runner.

Reply:
Totally agree that this may mean a bug in the SDK harness if it happens. I'm just thinking that a log message may be not enough as it's easy to be ignored. An exception makes sure that problem was discovered in time. But I'm fine if you think log is enough as users will investigate the log if there are serious problem for them after all. So, we need only add a WARNING log for this case, right?

Reply:
It is an error if we receive a response for something we weren't expecting.

We should close the control/data/state channel. The runner is then responsible for restarting the SDK harness and resuming processing.